omod/src/main/java/org/openmrs/module/reportingrest/web/controller/ReportingRestController.java

```diff
@@ -35,4 +54,94 @@ public class ReportingRestController extends MainResourceController {
+    @ResponseStatus(HttpStatus.OK)
+    public void runReport(@RequestBody RunReportRequestDTO runReportRequestDTO) {
+        ReportDefinition reportDefinition = Context.getService(ReportDefinitionService.class)
+            .getDefinitionByUuid(runReportRequestDTO.getReportDefinitionUuid());
+
+        ReportService reportService = getReportService();
+
+        Map<String, Object> parameterValues = new HashMap<String, Object>();
+        for (Parameter parameter : reportDefinition.getParameters()) {
+            Object convertedObj =
+                ConversionUtil.convert(runReportRequestDTO.getReportParameters().get(parameter.getName()), parameter.getType());
+            parameterValues.put(parameter.getName(), convertedObj);
+        }
+
+        List<RenderingMode> renderingModes = reportService.getRenderingModes(reportDefinition);
+        RenderingMode renderingMode = null;
+        for (RenderingMode mode : renderingModes) {
+            if (StringUtils.equals(mode.getArgument(), runReportRequestDTO.getRenderModeUuid())) {
+                renderingMode = mode;
+                break;
+            }
+        }
+
+        final ReportRequest reportRequest;
+        if (StringUtils.isNotBlank(runReportRequestDTO.getExistingRequestUuid())) {
+            reportRequest = reportService.getReportRequestByUuid(runReportRequestDTO.getExistingRequestUuid());
+        } else {
+            reportRequest = new ReportRequest();
+        }
+        reportRequest.setReportDefinition(new Mapped<ReportDefinition>(reportDefinition, parameterValues));
+        reportRequest.setRenderingMode(renderingMode);
+        reportRequest.setPriority(ReportRequest.Priority.NORMAL);
+        reportRequest.setSchedule(runReportRequestDTO.getSchedule());
+
+        reportService.queueReport(reportRequest);
+        reportService.processNextQueuedReports();
+    }
+
+    @RequestMapping(value = "/saveReport", method = RequestMethod.POST)
+    @ResponseStatus(HttpStatus.OK)
+    public void saveReport(@RequestParam String reportRequestUuid) {
+        ReportService reportService = getReportService();
+        ReportRequest reportRequest = reportService.getReportRequestByUuid(reportRequestUuid);
+        if (ReportRequest.Status.COMPLETED.equals(reportRequest.getStatus())) {
+            Report report = reportService.loadReport(reportRequest);
+            reportService.saveReport(report, StringUtils.EMPTY);
+        }
+    }
+
+    @RequestMapping(value = "/downloadReport", method = RequestMethod.GET)
+    @ResponseStatus(HttpStatus.OK)
+    public ReportFile downloadReport(@RequestParam String reportRequestUuid) {
+        return processAndDownloadReport(reportRequestUuid, getReportService());
+    }
+
+    @RequestMapping(value = "/downloadMultipleReports", method = RequestMethod.GET)
+    @ResponseStatus(HttpStatus.OK)
+    public List<ReportFile> downloadMultipleReports(@RequestParam String reportRequestUuids) {
```

Comment:
What's the practical use-case for this?

Reply:
It is needed for downloading multiple reports in our 3.X Reports pages. 
In this [PR](https://github.com/openmrs/openmrs-esm-admin-tools/pull/45) are screenshot attached. You can select multiple reports, then Download button appears and you can download them.

Reply:
Thanks for the feedback.